File: leetcode/Binary_Tree_Level_Order_Traversal_2.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int> > levelOrderBottom(TreeNode *root) {
        if(root == nullptr) return {}; 
        if(!root->left && !root->right) return {{root->val}};
     
        vector<vector<int>> result;
        queue<TreeNode *> p, q;
        p.push(root);
     
        vector<int> layer;
        while(!p.empty()){
            TreeNode *t = p.front();
            p.pop();
     
            layer.push_back(t->val);
            if(t->left) q.push(t->left);
            if(t->right) q.push(t->right);
     
            if(p.empty()){
                swap(p, q); 
                result.push_back(layer);
                layer.clear();
            }   
        }   
     
        reverse(result.begin(), result.end());
        return result;
    }
};

File: leetcode/java/BinaryTreeLevelOrderTraversal2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        if(root == null) return Collections.emptyList();
        Queue<TreeNode> q = new ArrayDeque<>();
        q.offer(root);
        
        Queue<TreeNode> p = new ArrayDeque<>();
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> layer = new ArrayList<>();
        
        while(!q.isEmpty()) {
            TreeNode t = q.poll();
            layer.add(t.val);
            if(t.left != null) p.offer(t.left);
            if(t.right != null) p.offer(t.right);
            if(q.isEmpty()) {
                Queue<TreeNode> k = q;
                q = p;
                p = k;
                result.add(new ArrayList<>(layer));
                layer.clear();
            }
        }
        Collections.reverse(result);
        return result;
    }
}
